subsys/net/ip/tcp.c

```diff
@@ -349,6 +364,25 @@ static void tcp_send(struct net_pkt *pkt)
 	tcp_pkt_unref(pkt);
 }
 
+static int tcp_derive_randomized_rto(int retry_count)
+{
+	uint32_t rand_val = sys_rand32_get();
+	int rto;
+
+	/* Compute a random value between 0 and
+	 * CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT << retry_count
+	 *
+	 * Every retransmit, the retransmission timeout increases by a factor 2
+	 */
+	rto = CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT << retry_count;
+	rto = rand_val % rto;
```

Comment:
Out of curiosity, is this backoff calculation algorithm covered by some RFC? I'm asking, as it seems a bit counter intuitive to me that every single retransmission timeout is randomized, and the random number is derived from a range (0, current_max), not some moving window. In result, it can turn out that the randomized backoff time for retransmission `n` is smaller than it was for `n - 1`.

For example in CoAP the RFC is clear about the retransmission time randomization - you only randomize the initial timeout, and then you double it on every retransmission.

Reply:
Nah, I didn't find any specifics either

Reply:
What do you suggest now?

Reply:
Well, w/o having backup in some actual document it's debatable what the best choice is (perhaps @jukkar you know some reliable backoff calculation mechanism backed by some RFC?). I think though that at least we should try to guarantee that conesutive timeout is **no smaller** than the previous one, it just sounds sane to me. So we can either:
* Follow CoAP here, and randomize only the initial timeout. This has a penalty in the memory consumption though, as you need to store the timeout value.
* Or simply specify `tmin = CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT << retry_count - 1` and `tmax = CONFIG_NET_TCP_INIT_RETRANSMISSION_TIMEOUT << retry_count` and choose a random number from `<tmin; tmax>` range.

Reply:
After quick search, Wikipedia has this article about this
https://en.wikipedia.org/wiki/Exponential_backoff
which has a link to SIP RFC
https://datatracker.ietf.org/doc/html/rfc3261
and there is discussion about exponential backoff. Not sure if that is useful here but perhaps can give some ideas.

Reply:
I took inspiration from the CoAP randomized retry structure to implement the randomized retransmit timeout. There is just a uint8_t added for each connection to maintain the random state.